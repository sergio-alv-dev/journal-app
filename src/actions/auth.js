import { types } from "../types/types";
import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { uiFinishLoading, uiStartLoading } from "./ui";
import Swal from 'sweetalert2';
import { notesLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    
    
    return ( dispatch ) => {
        
        dispatch( uiStartLoading() );

         return firebase.auth().signInWithEmailAndPassword( email, password )
            .then( ({ user }) => {
                // console.log(user);
                dispatch( login(user.uid, user.displayName ) );
                dispatch( uiFinishLoading() );
            })
            .catch( e => {
                console.log(e);
                dispatch( uiFinishLoading() );
                Swal.fire('Error', e.message, 'error');
            })
        

    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return( dispatch ) => {
        firebase.auth().createUserWithEmailAndPassword( email, password )
        .then( async ({ user }) => {
            
            await user.updateProfile({ displayName:name });
            console.log(user);
            dispatch(
                login(user.uid, user.displayName )
            )
            
        })
        .catch( e => {
            console.log(e)
            Swal.fire('Error', e.message, 'error');
        })
    }

}

export const startGoogleLogin = () => {
    return ( dispatch ) => {

        // console.log( 'google' );

        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch(
                    login(user.uid, user.displayName )
                )
                // console.log( userCred );
            })
            .catch( e => {
                console.log(e)
                Swal.fire('Error', e.message, 'error');

            })

    }
}


export const login = ( uid, displayName) => {
    return{
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const startLogout = () => {
    return async ( dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() );
        dispatch( notesLogout() );
    }
}

export const logout = () => ({
    type: types.logout
})