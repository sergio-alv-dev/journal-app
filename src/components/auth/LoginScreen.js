import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { loading } = useSelector( state => state.ui);

    const [ formValues, handleInputChange ] = useForm({
        email: '',
        password: ''
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if ( isFormValid() ){
            dispatch( startLoginEmailPassword( email, password ) );
        }
    }

    const handleGoogleLogin = () => { 
        // console.log('click handleGoogleLogin')
        dispatch ( startGoogleLogin() );
    };

    const isFormValid = () => {

        if ( !validator.isEmail( email ) ) {
            dispatch( setError('Email is not valid') );
            return false;
        } else if ( password.length < 5 ) {
            dispatch( setError('Password is not valid') );
            return false;
        }
            
        dispatch( removeError() );
        
        return true;
    }

    return (
        <>
            <div
                className = "animate__animated animate__fadeIn"    
                >

            <h3 className= "auth__title mb-5"> Login </h3>

            <form 
                onSubmit = { handleLogin }

            >
                <input 
                    type = "text"
                    placeholder = "Email"
                    name = "email"
                    className = "auth__input"
                    autoComplete = "off"
                    value = { email }
                    onChange = { handleInputChange }
                />
                
                <input 
                    type = "password"
                    placeholder = "Password"
                    name = "password"
                    className = "auth__input"
                    value = { password }
                    onChange = { handleInputChange }
                />

                <button
                    type = "submit"
                    className = "btn btn-primary btn-block"
                    disabled = { loading }
                >
                    Login
                </button>
                
                <div className = "auth__social-networks mb-1">
                    <p>Login with Social networks</p>

                    <div 
                        className="google-btn" 
                        onClick = { handleGoogleLogin }
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <hr/>
                <Link 
                    to = "/auth/register"
                    className = "link"    
                >
                Create new Account
                </Link>
            </form>
        </div>

        </>
    )
}
