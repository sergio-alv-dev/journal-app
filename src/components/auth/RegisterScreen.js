import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';

import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

// se puede utilizar useState, no es necesario utilizar redux aqui
    const dispatch = useDispatch();
    const { loading, msgError } = useSelector( state => state.ui );
    
    const [ formValues, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });
    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        
        if ( isFormValid() ) {
            dispatch( startRegisterWithEmailPasswordName( email, password, name ));
        }
    }

    const isFormValid = () => {

        if( name.trim().length === 0){
            dispatch( setError('Name is required') );
            return false;
        } else if ( !validator.isEmail( email ) ) {
            dispatch( setError('Email is not valid') );
            return false;
        } else if ( password !== password2 || password.length < 5 ) {
            dispatch( setError('Password is not valid') );
            return false;
        } else {
            dispatch( removeError() );
        }

        return true;
    }



    return (
        <>
        < div
                className = "animate__animated animate__fadeIn"    >

            
            <h3 className= "auth__title mb-5"> Register </h3>
            
            {
                msgError 
                &&
                (
                    <div className="auth__alert-error">
                        {
                            msgError 
                        }
                    </div>
                )
            }

            <form onSubmit = { handleRegister }>

                <input 
                    type = "text"
                    placeholder = "Name"
                    name = "name"
                    className = "auth__input"
                    autoComplete = "off"
                    value = { name }
                    onChange = { handleInputChange }
                />
                <input 
                    type = "text"
                    placeholder = "Email"
                    name = "email"
                    className = "auth__input"
                    autoComplete = "off"
                    value = { email }
                    onChange = { handleInputChange }
                />
                
                <input 
                    type = "password"
                    placeholder = "Password"
                    name = "password"
                    className = "auth__input"
                    value = { password }
                    onChange = { handleInputChange }
                />

                <input 
                    type = "password"
                    placeholder = "Confirm Password"
                    name = "password2"
                    className = "auth__input"
                    value = { password2 }
                    onChange = { handleInputChange }
                />

                <button
                    type = "submit"
                    className = "btn btn-primary btn-block mb-5"
                    disabled = { loading }
                >
                    Register
                </button>
                
                <hr/>
                <Link
                    to = "/auth/login"
                    className = "link mt-5"    
                >
                    Already registered?
                </Link>
            </form>
            </div>

        </>
    )
}
