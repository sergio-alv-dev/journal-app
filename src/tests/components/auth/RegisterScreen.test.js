import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';


jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}))

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore( initState )
// store.dispatch = jest.fn();

const wrapper = mount( 
    <MemoryRouter>
        <Provider store = { store }>
            <RegisterScreen />
        </Provider>
    </MemoryRouter>
)


describe('Pruebas en <RegisterScreen />', () => {

    test('debe de hacer match con el Snapshot', () => {

        expect( wrapper ).toMatchSnapshot();
        
    });

    test('debe de hacer el dispatch de la acciÃ³n respectiva', () => {
        
        const emailField = wrapper.find('input[name="email"]');

        // console.log(emailField.exists());
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();

        expect( actions[0] ).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        }); 
    });

    test('debe de mostrar la caja de alerta con el error', () => {

        
    const initState = {
        auth: {},
        ui: {
            loading: false,
            msgError: 'Email no es correcto',
        }
    };

    const store = mockStore( initState )

    const wrapper = mount( 
        <Provider store = { store }>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    );

    expect( wrapper.find('.auth__alert-error').exists() ).toBe(true);
    expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );

        
    })
    
    
    
    
})
