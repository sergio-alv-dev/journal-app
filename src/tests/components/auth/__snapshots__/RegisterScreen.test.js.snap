// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> debe de hacer match con el Snapshot 1`] = `
<div
  className="animate__animated animate__fadeIn"
>
  <h3
    className="auth__title mb-5"
  >
     Register 
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="chechitooo"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="chechox5@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <hr />
    <a
      className="link mt-5"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
